File: desktop/src/main/announcement-service.ts
import fs from 'fs';
import path from 'path';
import os from 'os';
import { log } from './logger';
import { isExpired, type Announcement } from '../shared/announcement';

/**
 * Announcement Service
 *
 * Fetches the first non-comment line from youcoded's announcements.txt
 * (repo root) and caches it at ~/.claude/.announcement-cache.json for the
 * status bar widget and the terminal statusline.
 *
 * Format on the wire (first non-# line wins):
 *   YYYY-MM-DD: message         -> expires at that date
 *   message                     -> never expires
 *   (empty or comments only)    -> no announcement
 *
 * Lifecycle bugs this service is responsible for preventing:
 *   - Fetch-time expiry filter: past-date lines are treated as empty so
 *     already-stale content never lands in the cache.
 *   - Clear propagation: when the remote file is empty, we *write* a
 *     cleared cache ({ message: null, fetched_at }) rather than returning
 *     without touching disk. This is what lets the status bar pill
 *     disappear within the refresh interval instead of lingering.
 *
 * Called on app launch and every 1h while running.
 */

const ANNOUNCEMENTS_URL =
  'https://raw.githubusercontent.com/itsdestin/youcoded/master/announcements.txt';
const CACHE_PATH = path.join(os.homedir(), '.claude', '.announcement-cache.json');
const TMP_PATH = `${CACHE_PATH}.tmp`;
const REFRESH_MS = 60 * 60 * 1000; // 1 hour

function writeAtomic(data: Announcement): void {
  fs.mkdirSync(path.dirname(CACHE_PATH), { recursive: true });
  try {
    fs.writeFileSync(TMP_PATH, JSON.stringify(data, null, 2), 'utf8');
    fs.renameSync(TMP_PATH, CACHE_PATH);
  } catch (e) {
    try { fs.unlinkSync(TMP_PATH); } catch { /* best-effort cleanup */ }
    throw e;
  }
}

// Returns { message, expires? } for a valid unexpired line, or null for
// empty/comments-only/expired input. Callers treat null as "clear the cache."
function parseAnnouncement(text: string): { message: string; expires?: string } | null {
  for (const line of text.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;
    const dateMatch = trimmed.match(/^(\d{4}-\d{2}-\d{2}): (.+)$/);
    if (dateMatch) {
      const expires = dateMatch[1];
      // Fetch-time expiry filter: already-past dates are treated as empty.
      // Prevents a freshly-fetched stale entry from ever entering the cache.
      if (isExpired(expires)) return null;
      return { message: dateMatch[2].trim(), expires };
    }
    return { message: trimmed };
  }
  return null;
}

/**
 * Fetch announcements once. Network failure or non-200 leaves the existing
 * cache intact (users don't lose announcements on going offline). An empty
 * or all-expired remote file writes a cleared cache so the status bar
 * pill disappears within the refresh interval.
 */
export async function fetchAnnouncement(): Promise<void> {
  let response: Response;
  try {
    response = await fetch(ANNOUNCEMENTS_URL);
  } catch {
    return; // offline / DNS failure — leave cache alone
  }
  if (!response.ok) return;

  const text = await response.text();
  const parsed = parseAnnouncement(text);

  const cache: Announcement = parsed
    ? {
        message: parsed.message,
        fetched_at: new Date().toISOString(),
        ...(parsed.expires ? { expires: parsed.expires } : {}),
      }
    : {
        // Clear-propagation write. null (vs. undefined) distinguishes
        // "explicitly cleared" from "no cache file yet" on the reader side.
        message: null,
        fetched_at: new Date().toISOString(),
      };

  try {
    writeAtomic(cache);
    log('INFO', 'Announcements', 'Cache updated', {
      cleared: cache.message === null,
      hasExpiry: !!cache.expires,
    });
  } catch (e) {
    log('ERROR', 'Announcements', 'Cache write failed', { error: String(e) });
  }
}

let refreshTimer: NodeJS.Timeout | null = null;

/**
 * Kick off the first fetch immediately and schedule a 1h refresh loop. Safe
 * to call more than once; subsequent calls reset the timer.
 */
export function startAnnouncementService(): void {
  stopAnnouncementService();
  fetchAnnouncement().catch(e =>
    log('ERROR', 'Announcements', 'Initial fetch threw', { error: String(e) }),
  );
  refreshTimer = setInterval(() => {
    fetchAnnouncement().catch(e =>
      log('ERROR', 'Announcements', 'Scheduled fetch threw', { error: String(e) }),
    );
  }, REFRESH_MS);
  // Don't keep the event loop alive just for this — app shutdown shouldn't wait.
  refreshTimer.unref?.();
}

export function stopAnnouncementService(): void {
  if (refreshTimer) {
    clearInterval(refreshTimer);
    refreshTimer = null;
  }
}

// Exposed for tests
export const __test = { parseAnnouncement };

File: app/src/main/kotlin/com/youcoded/app/runtime/AnnouncementService.kt
package com.youcoded.app.runtime

import android.util.Log
import org.json.JSONObject
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/**
 * Android-side announcement fetcher. Mirror of desktop's
 * `announcement-service.ts`. Fetches youcoded/announcements.txt once per
 * hour and writes ~/.claude/.announcement-cache.json so that:
 *
 *   - the terminal statusline.sh can render the ★ announcement line, AND
 *   - SessionService.startStatusBroadcast() can fold the cache into
 *     status:data for the React status-bar widget.
 *
 * Lifecycle rules (must match desktop):
 *   - Fetch-time expiry filter: if the first non-comment line has a
 *     YYYY-MM-DD prefix that is strictly less than today (device local
 *     date), treat it as empty. Already-stale content never reaches cache.
 *   - Clear propagation: when the remote file is empty (after comment
 *     stripping), write { message: null, fetched_at } so readers can
 *     distinguish "explicitly cleared" from "no cache yet."
 *   - Offline tolerance: any exception during fetch leaves the existing
 *     cache untouched.
 *
 * Launch from SessionService.onCreate() via startAnnouncementService(homeDir).
 */
class AnnouncementService(private val homeDir: File) {

    private var timer: java.util.Timer? = null

    fun start() {
        stop()
        timer = java.util.Timer("announcement-fetch", true).apply {
            scheduleAtFixedRate(object : java.util.TimerTask() {
                override fun run() {
                    try {
                        fetchOnce()
                    } catch (e: Exception) {
                        Log.w(TAG, "fetch threw: ${e.message}")
                    }
                }
            }, INITIAL_DELAY_MS, REFRESH_MS)
        }
    }

    fun stop() {
        timer?.cancel()
        timer = null
    }

    private fun fetchOnce() {
        val text: String = try {
            val conn = (URL(URL_STR).openConnection() as HttpURLConnection).apply {
                connectTimeout = 10_000
                readTimeout = 10_000
                requestMethod = "GET"
            }
            try {
                if (conn.responseCode !in 200..299) return
                conn.inputStream.bufferedReader().use { it.readText() }
            } finally {
                conn.disconnect()
            }
        } catch (e: Exception) {
            // Offline / DNS / network: leave existing cache alone.
            return
        }

        val parsed: ParsedAnnouncement? = parseAnnouncement(text)

        val nowIso = iso8601Now()
        val json = JSONObject()
        if (parsed != null) {
            json.put("message", parsed.message)
            json.put("fetched_at", nowIso)
            if (parsed.expires != null) json.put("expires", parsed.expires)
        } else {
            // Explicit clear: message: null lets StatusBar's isExpired gate
            // hide the pill even if the on-disk cache was previously
            // populated with an unrelated entry.
            json.put("message", JSONObject.NULL)
            json.put("fetched_at", nowIso)
        }

        writeAtomic(json.toString(2))
    }

    private fun writeAtomic(contents: String) {
        val claudeDir = File(homeDir, ".claude").apply { mkdirs() }
        val cache = File(claudeDir, ".announcement-cache.json")
        val tmp = File(claudeDir, ".announcement-cache.json.tmp")
        try {
            tmp.writeText(contents, Charsets.UTF_8)
            if (!tmp.renameTo(cache)) {
                // Rename can fail if dest is on a different mount; fall back
                // to copy + delete to preserve atomic-ish behavior.
                cache.writeText(contents, Charsets.UTF_8)
                tmp.delete()
            }
        } catch (e: Exception) {
            try { tmp.delete() } catch (_: Exception) {}
            Log.w(TAG, "cache write failed: ${e.message}")
        }
    }

    companion object {
        private const val TAG = "AnnouncementService"
        private const val URL_STR =
            "https://raw.githubusercontent.com/itsdestin/youcoded/master/announcements.txt"
        private const val INITIAL_DELAY_MS = 5_000L
        private const val REFRESH_MS = 60L * 60L * 1000L // 1 hour

        data class ParsedAnnouncement(val message: String, val expires: String?)

        // Pure function; unit-testable in isolation if needed.
        fun parseAnnouncement(text: String): ParsedAnnouncement? {
            val datePrefix = Regex("""^(\d{4}-\d{2}-\d{2}): (.+)$""")
            for (raw in text.split('\n')) {
                val trimmed = raw.trim()
                if (trimmed.isEmpty() || trimmed.startsWith("#")) continue
                val m = datePrefix.matchEntire(trimmed)
                if (m != null) {
                    val expires = m.groupValues[1]
                    // Fetch-time expiry filter — mirrors desktop's isExpired.
                    if (expires < todayYYYYMMDD()) return null
                    return ParsedAnnouncement(m.groupValues[2].trim(), expires)
                }
                return ParsedAnnouncement(trimmed, null)
            }
            return null
        }

        private fun todayYYYYMMDD(): String {
            val fmt = SimpleDateFormat("yyyy-MM-dd", Locale.US).apply {
                timeZone = TimeZone.getDefault()
            }
            return fmt.format(Date())
        }

        private fun iso8601Now(): String {
            val fmt = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
                timeZone = TimeZone.getTimeZone("UTC")
            }
            return fmt.format(Date())
        }
    }
}
